Tests/NIOSSLTests/SSLContextTests.swift

```diff
@@ -0,0 +1,110 @@
+        let (cert1, key1) = generateSelfSignedCert()
+        SSLContextTest.cert1 = cert1
+        SSLContextTest.key1 = key1
+        let (cert2, key2) = generateSelfSignedCert()
+        SSLContextTest.cert2 = cert2
+        SSLContextTest.key2 = key2
+    }
+
+    private func configuredClientSSLContext() throws -> NIOSSLContext {
+        var config = TLSConfiguration.makeServerConfiguration(
+            certificateChain: [.certificate(SSLContextTest.cert2)],
+            privateKey: .privateKey(SSLContextTest.key2)
+        )
+        config.trustRoots = .certificates([SSLContextTest.cert2])
+        let context = try NIOSSLContext(configuration: config)
+        return context
+    }
+
+    private func configuredServerSSLContext(eventLoop: EventLoop) throws -> NIOSSLContext {
+        // Initialize with cert1
+        var config = TLSConfiguration.makeServerConfiguration(
+            certificateChain: [.certificate(SSLContextTest.cert1)],
+            privateKey: .privateKey(SSLContextTest.key1)
+        )
+        // Configure callback to return cert2
+        config.sslContextCallback = { (values, _) in
+            let alternateConfig = TLSConfiguration.makeServerConfiguration(
+                certificateChain: [.certificate(SSLContextTest.cert2)],
+                privateKey: .privateKey(SSLContextTest.key2)
+            )
+            let alternateContext = try NIOSSLContext(configuration: alternateConfig)
+            return eventLoop.makeSucceededFuture(alternateContext)
+        }
+        return try NIOSSLContext(configuration: config)
+    }
+
+    private func assertSniResult(sniField: String?, expectedResult: String) throws {
+        let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        defer {
+            try? group.syncShutdownGracefully()
+        }
+
+        let handshakeResultPromise = group.next().makePromise(of: Void.self)
+        let handshakeWatcher = WaitForHandshakeHandler(handshakeResultPromise: handshakeResultPromise)
+
+        let clientContext = try configuredClientSSLContext()
+        let serverContext = try configuredServerSSLContext(eventLoop: group.next())
+
+        let serverChannel = try serverTLSChannel(context: serverContext, preHandlers: [], postHandlers: [], group: group)
+        defer {
+            _ = try? serverChannel.close().wait()
+        }
+
+        let clientChannel = try clientTLSChannel(context: clientContext,
+                                                 preHandlers: [],
+                                                 postHandlers: [handshakeWatcher],
+                                                 group: group,
+                                                 connectingTo: serverChannel.localAddress!,
+                                                 serverHostname: sniField)
+        defer {
+            _ = try? clientChannel.close().wait()
+        }
+
+        try handshakeResultPromise.futureResult.wait()
+        XCTAssertEqual(true, true)
```

Comment:
We're never getting here, still debugging. The new cert_cb is correctly going through the two phases and returning `1` in the 2nd case, but the handshake is not completing.